### ISSUE TITLE
[BUG] AQE does not respect `entirePlanWillNotWork`

### ISSUE THREAD

**Issue by andygrove** [NVIDIA/spark-rapids#6230]
**Describe the bug**
In `com.nvidia.spark.rapids.RapidsMeta#entirePlanWillNotWork` we do not tag the underlying plan via `willNotWorkOnGpu` so when AQE is preparing query stages, it is possible that some parts of the plan will run on GPU and others stay on CPU and this can result in an invalid plan that fails to execute.

**To Reproduce**
I found this while attempting to make more Delta Lake metadata queries fall back to CPU. Here is an example:

```
Transformed Plan:
Exchange hashpartitioning(_row_id_#2100L, 200), ENSURE_REQUIREMENTS, [id=#1845]
+- HashAggregate(keys=[_row_id_#2100L], functions=[partial_sum(one#2121)], output=[_row_id_#2100L, sum#2206L])
   +- Project [_row_id_#2100L, UDF(_file_name_#2104) AS one#2121]
      +- BroadcastHashJoin [c0#2020], [c0#2016], Inner, BuildLeft, false
         :- ColumnarToRow
         :  +- BroadcastQueryStage 0
         :     +- GpuBroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true]),false), [id=#1717]
         :        +- GpuProject [c0#2020]
         :           +- GpuCoalesceBatches targetsize(2147483647)
         :              +- GpuFilter gpuisnotnull(c0#2020), true
         :                 +- GpuRowToColumnar targetsize(2147483647)
         :                    +- *(1) Filter UDF()
         :                       +- GpuColumnarToRow false
         :                          +- GpuFileGpuScan parquet [c0#2020,c1#2021] Batched: true, DataFilters: [], Format: Parquet, Location: PreparedDeltaFileIndex[file:/tmp/t2], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c0:string,c1:int>
         +- Filter isnotnull(c0#2016)
            +- Project [c0#2016, _row_id_#2100L, input_file_name() AS _file_name_#2104]
               +- Project [c0#2016, monotonically_increasing_id() AS _row_id_#2100L]
                  +- ColumnarToRow
                     +- FileScan parquet [c0#2016] Batched: true, DataFilters: [], Format: Parquet, Location: TahoeBatchFileIndex(1 paths)[file:/tmp/t1], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c0:string>
```

This fails with the following error:

```
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:208)
  at org.apache.spark.sql.execution.ColumnarToRowExec.<init>(Columnar.scala:69)
```

**Expected behavior**
We should recursively tag the plan using `willNotWorkOnGpu` so that AQE works correctly.

**Additional context**
None

**Comment by andygrove** [comment]
Here is one way of fixing this. I will create a PR once I have a suitable repro test case.

```
   final def entirePlanWillNotWork(because: String): Unit = {
     cannotReplaceAnyOfPlanReasons.get.add(because)
+    // recursively tag the plan so that AQE respects this
+    willNotWorkOnGpu(because)
+    childPlans.foreach(_.entirePlanWillNotWork(because))
   }
```

**Comment by sameerz** [comment]
Closed by #6250
